docs/design-approaches.md

```diff
@@ -0,0 +1,235 @@
+2. Orchestrator needs the VM handle in some way. This is required for:
+   - Invoking the firecracker MMDS API (for updating IAM Role credentials and metadata)
+   - Draining VM log and metrics fifos
+3. A VM should not be terminated as long as there are some containers running in it (unless it's an impairment event like reboot)
+
+
+## Glossary
+
+* Containerd - OCI compatible container runtime. Receives container lifecycle/management requests and performs them using configured plugins/options.
+* Runtime - An implementation of the runtime v2 [spec](https://github.com/containerd/containerd/tree/master/runtime/v2) for containerd. Unlike the default implementation which runs runc directly, our runtime will be responsible for managing Firecracker instances as well as communicating container requests to the inside-the-vm-agent.
+* Agent - A process inside of the running VM. Manages containers, handles communication with the Runtime  and utilizes runc to launch containers inside the microVM.
+* Orchestrator - An entity that manages tasks and submits requests to lower level subsystems. This is where all the decisions should be made as it is the piece with the largest amount of information about the container-group execution needs. It's where we can decide vm size, etc.
+* VMID - abstract identifier not related to any specific implementation. Means a group of containers somehow related by same id (for example a group of containers from pod or task definition)
+
+
+## High-level overview
+
+![design-overview](img/design-overview.png)
+
+1. Orchestrator accepts task definition request from data-plane and breaks it into sub-requests suitable for sending to containerd (using `CreateTaskRequest` [struct](https://github.com/containerd/containerd/blob/master/api/services/tasks/v1/tasks.pb.go#L79), additional parameters can be passed to runtime though `Options` [field](https://github.com/containerd/containerd/blob/master/api/services/tasks/v1/tasks.pb.go#L93))
+2. containerd uses firecracker-containerd runtime and block device snapshotter. Requests are passed to runtime. Runtime is responsible for communicating with Agent, attaching/managing block devices and controlling container execution lifecycle.
+3. Runtime controls the list of drives to use for passing container mounts (more on this in the following sections).
+4. Runtime and/or FC-control [finds](https://github.com/firecracker-microvm/firecracker-containerd/blob/7fd9dcfb77c6b5e760ed0a8b2e1eecf44fe37c20/runtime/service.go#L560) available context [identifier](http://man7.org/linux/man-pages/man7/vsock.7.html) for Vsock and starts the VM (unless it's already started), waits for the VM to boot up and Agent to start responding to requests via vsock channel. After that the Runtime submits requests for managing containers through Agent (interface is described below).
```

Comment:
I think finding a CID is not high level and is subject to change depending on how Firecracker implements vsock.